pkg/admissioncontroller/server/handlers/webhooks/validate_resource_size.go

```diff
@@ -0,0 +1,171 @@
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apiserver/pkg/authentication/serviceaccount"
+)
+
+type objectSizeHandler struct {
+	config         *apisconfig.ResourceAdmissionConfiguration
+	codecs         serializer.CodecFactory
+	logger         logrus.FieldLogger
+	maxRequestBody int64
+}
+
+const (
+	validatorName = "resource_size_validator"
+)
+
+// NewValidateResourceSizeHandler creates a new handler for validating the resource size of a request.
+func NewValidateResourceSizeHandler(config *apisconfig.ResourceAdmissionConfiguration) http.HandlerFunc {
+	scheme := runtime.NewScheme()
+	utilruntime.Must(admissionregistrationv1beta1.AddToScheme(scheme))
+
+	h := &objectSizeHandler{
+		config: config,
+		codecs: serializer.NewCodecFactory(scheme),
+		logger: logger.NewFieldLogger(logger.Logger, "component", validatorName),
+
+		// Take the same, fixed value from API server for general safety to reduce the odds of OOM issues while reading the body.
+		// https://github.com/kubernetes/kubernetes/blob/d8eac8df28e6b50cd0f5380e23fc57daaf92972e/staging/src/k8s.io/apiserver/pkg/server/config.go#L322
+		maxRequestBody: int64(3 * 1024 * 1024),
+	}
+	return h.ValidateResourceSize
+}
+
+// ValidateResourceSize is a HTTP handler for validating whether the incoming resource does not exceed the configured limit.
+func (h *objectSizeHandler) ValidateResourceSize(w http.ResponseWriter, r *http.Request) {
+	var (
+		deserializer   = h.codecs.UniversalDeserializer()
+		receivedReview = &admissionv1beta1.AdmissionReview{}
+	)
+
+	if err := DecodeAdmissionRequest(r, deserializer, receivedReview, h.maxRequestBody); err != nil {
+		h.logger.Errorf(err.Error())
+		respond(w, errToAdmissionResponse(err))
+		return
+	}
+
+	logEntry := h.logger.WithField("resource", receivedReview.Request.Resource).WithField("name", receivedReview.Request.Name)
+	if receivedReview.Request.Namespace != "" {
+		logEntry = logEntry.WithField("namespace", receivedReview.Request.Namespace)
+	}
+
+	// Now that all checks have been passed we can actually validate the admission request.
+	reviewResponse := h.admit(receivedReview.Request, logEntry)
+	if !reviewResponse.Allowed && reviewResponse.Result != nil {
+		logEntry.Infof("Rejected request of user '%s': %v", receivedReview.Request.UserInfo.Username, reviewResponse.Result.Message)
+	}
+	respond(w, reviewResponse)
+}
+
+// admit does only allow the request if the object in the request does not exceed a configured limit.
+func (h *objectSizeHandler) admit(request *admissionv1beta1.AdmissionRequest, logEntry logrus.FieldLogger) *admissionv1beta1.AdmissionResponse {
+	if request.SubResource != "" {
+		return admissionResponse(true, "")
+	}
+
+	if isUnrestrictedUser(request.UserInfo, h.config.UnrestrictedSubjects) {
```

Comment:
This would panic with nil pointer if `resourceAdmissionConfiguration` is not configured in the component config. We could fail early at component start that `resourceAdmissionConfiguration` is required or we can add nil check.

Reply:
Nice catch! I now default the configuration if its not set.